### PR TITLE
Add support for events, programs, and kernel submission to dpctl.

### PR DESCRIPTION
**Pull request by diptorupd**
* New dpctl C API wrappers for sycl::event, sycl::kernel and sycl::program.
    * Wrappers to create an OpenCL interoperability program from either string
      or SPIR-V IL files.
    * Support for submitting an interoperability kernel to a sycl::queue. We
      support using either a range or an nd_range to submit the kernel.
    * Event wait function is now exposed via the C API.
    * New test cases for the C API.
    * The dpctl Cython wrapper now supports program creation and kernel
      submission.
    * Other refactoring to standardize function names, attribute names, comments.

### CODE REVIEW

**Comment by diptorupd** [comment]
Closes #52, #77.

File: dpctl/backend.pxd [hunk 1]
```diff
@@ -67,27 +76,86 @@ cdef extern from "dppl_sycl_device_interface.h":
...
+                                const size_t gRange[3],
+                                const size_t lRange[3],
+                                size_t NDims,
+                                const DPPLSyclEventRef *DepEvents,
+                                size_t NDepEvents)
+    cdef void DPPLQueue_Wait (const DPPLSyclQueueRef QRef)
     cdef void DPPLQueue_Memcpy (const DPPLSyclQueueRef Q,
-                                void *Dest, const void *Src, size_t Count) \
-         except +
+                                void *Dest, const void *Src, size_t Count)
```
Comment (PokhodenkoSA) [inline dpctl/backend.pxd:155]: @diptorupd as I remember you wanted to make Memcpy to return SyclEvent. In the next PR?

File: dpctl/tests/__init__.py [hunk 2]
```diff
@@ -22,6 +22,8 @@
 ## Top-level module of all dpctl Python unit test cases.
 ##===----------------------------------------------------------------------===##
 
+from .test_dump_functions import *
+from .test_sycl_kernel_submit import *
+from .test_sycl_program import *
 from .test_sycl_queue_manager import *
```
Comment (PokhodenkoSA) [inline dpctl/tests/__init__.py:28]:
```suggestion
from .test_sycl_queue_manager import *
from .test_sycl_queue_memcpy import *
```

File: dpctl/tests/test_sycl_kernel_submit.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            args.append(a.base)
+            args.append(b.base)
+            args.append(c.base)
+            args.append(ctypes.c_int(d))
+
+            r = [1024]
+
+            q.submit(axpyKernel, args, r)
+            self.assertTrue(np.allclose(c, a*d + b))
+
```
Comment (PokhodenkoSA) [inline dpctl/tests/test_sycl_kernel_submit.py:66]:
```suggestion


```

File: dpctl/tests/test_sycl_program.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+##
+## \file
+## Defines unit test cases for the SyclProgram and SyclKernel classes defined
+##  in sycl_core.pyx.
+##===----------------------------------------------------------------------===##
+
+import dpctl
+import unittest
+import os
+
```
Comment (PokhodenkoSA) [inline dpctl/tests/test_sycl_program.py:29]:
```suggestion


```

**Review by PokhodenkoSA — APPROVED** [review]
LGTM.

File: backends/include/dppl_sycl_program_interface.h [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+#include "Support/DllExport.h"
+#include "Support/ExternC.h"
+#include "Support/MemOwnershipAttrs.h"
+
+DPPL_C_EXTERN_C_BEGIN
+
+/*!
+ * @brief Create a Sycl program from an OpenCL SPIR-V binary file.
+ *
+ * Sycl 1.2 does expose any method to create a sycl::program from a SPIR-V IL
```
Comment (oleksandr-pavlyk) [inline backends/include/dppl_sycl_program_interface.h:43]: Did you mean 'does not expose'?

File: backends/include/dppl_sycl_program_interface.h [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+ * file. To get around this limitation, we need to use the Sycl feature to
+ * create an interoperability kernel from an OpenCL kernel. This function first
+ * creates an OpenCL program and kernel from the SPIR-V binary and then using
+ * the Sycl-OpenCL interoperability feature creates a Sycl kernel from the
+ * OpenCL kernel.
+ *
+ * The feature to create a Sycl kernel from a SPIR-V IL binary will be available
+ * in Sycl 2.0, at which point we may deprecate this function.
```
Comment (oleksandr-pavlyk) [inline backends/include/dppl_sycl_program_interface.h:51]: It is better to say, "at which point this function may become deprecated."

File: backends/source/dppl_sycl_queue_interface.cpp [hunk 7]
```diff
@@ -34,32 +34,230 @@ using namespace cl::sycl;
...
+DEFINE_SIMPLE_CONVERSION_FUNCTIONS(kernel, DPPLSyclKernelRef)
+DEFINE_SIMPLE_CONVERSION_FUNCTIONS(queue, DPPLSyclQueueRef)
+
+/*!
+ * @brief Set the kernel arg object
+ *
+ * @param    cgh            My Param doc
+ * @param    Arg            My Param doc
+ */
+bool set_kernel_arg (handler &cgh, size_t idx, __dppl_keep void *Arg,
```
Comment (oleksandr-pavlyk) [inline backends/source/dppl_sycl_queue_interface.cpp:49]: We will eventually need to support complexes as well. 

This is not straightforward with `ctypes`: https://stackoverflow.com/questions/13373291/complex-number-in-ctypes

So down the road we might just introduce our own scalar type in `dpctl`.

File: dpctl/backend.pxd (hunk 1 above)
Comment (diptorupd) [reply dpctl/backend.pxd:155]: Yes, I will open a follow up. Refer #84

File: backends/include/dppl_sycl_program_interface.h (hunk 5 above)
Comment (diptorupd) [reply backends/include/dppl_sycl_program_interface.h]: good catch. Fixed now.

File: backends/include/dppl_sycl_program_interface.h (hunk 6 above)
Comment (diptorupd) [reply backends/include/dppl_sycl_program_interface.h]: I changed the comment. Thanks.

File: backends/source/dppl_sycl_queue_interface.cpp (hunk 7 above)
Comment (diptorupd) [reply backends/source/dppl_sycl_queue_interface.cpp:49]: Opened #85 to track the issue.

**Pull request merged**